File: tinyos-0.4.x/tos/tools/expInf/Experiment.java
/*									tab:4
 * Experiment.java
 *
 * "Copyright (c) 2000 and The Regents of the University 
 * of California.  All rights reserved.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose, without fee, and without written agreement is
 * hereby granted, provided that the above copyright notice and the following
 * two paragraphs appear in all copies of this software.
 * 
 * IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY FOR
 * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT
 * OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF THE UNIVERSITY OF
 * CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
 * ON AN "AS IS" BASIS, AND THE UNIVERSITY OF CALIFORNIA HAS NO OBLIGATION TO
 * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS."
 *
 * Authors:  Solomon Bien
 *
 */


package expInf;

/**
 * This class provides an interface to the AM handlers of the EXPERIMENT 
 * component.  It allows the user to start and stop experiments, set the 
 * experimental ID, start and stop the topology_query component, and 
 * query a node for its information.
 */
public class Experiment extends ExpInfComp implements AMHandler {
    /**
     * AM handler to set an experimental ID
     */
    public static final byte AM_SET_EXP_ID_MSG = 92;
    /**
     * AM handler to query a node for its information
     */
    public static final byte AM_DISCOVER_MSG = 91;
    /**
     * AM handler to start an experiment
     */
    public static final byte AM_START_EXP_MSG = 90;
    /** 
     * AM handler to stop an experiment
     */
    public static final byte AM_STOP_EXP_MSG = 89;
    
    /**
     * UART address
     */
    public static final short GENERIC_BASE_ADDR = 0x7e;

    /**
     * Constructor takes as an argument an AMInterface object that is already 
     * open.
     */
    public Experiment(AMInterface aif) {
	super(aif);
	super.aif.registerHandler(this,AM_DISCOVER_MSG);
    }

    /**
     * Changes the experimental ID of node <nodeID> to <newExpID>
     */
    public void setExpID(byte newExpID, short nodeID) throws java.io.IOException {
	byte [] data = new byte[super.aif.AM_SIZE];
	
	data[0] = newExpID;
	
	super.aif.sendAM(data,AM_SET_EXP_ID_MSG,nodeID);
    }

    public void handleAM(byte[] data, short addr, byte id, byte group) {
	for(int i = 0;i<data.length;i++) {
	    System.out.print(data[i] + " ");
	}
	System.out.println("");
    }

    /**
     * Queries a node <nodeID> for its information
     */
    public void discover(short nodeID) throws java.io.IOException {
	byte [] data = new byte[super.aif.AM_SIZE];
	
	//&&&&&&&&&&&&CHANGE THIS&&&&&&&&
	//data[0] = (byte) (GENERIC_BASE_ADDR & 0xff);
	//data[1] = (byte) ((GENERIC_BASE_ADDR >> 8) & 0xff);	

	data[0] = (byte) (AMInterface.TOS_BCAST_ADDR & 0xff);
	data[1] = (byte) ((AMInterface.TOS_BCAST_ADDR >> 8) & 0xff);

	super.aif.sendAM(data,AM_DISCOVER_MSG,nodeID);
    }

    /**
     * Starts the experiment on node <nodeID> in <delay> clock ticks and 
     * runs it for <duration> clock ticks.  If <startConnectivity> is 1, 
     * it starts topology_query.  If 0, the loaded experiment is run.
     */
    private void startExp(byte startConnectivity, byte delay, byte duration, short nodeID) throws java.io.IOException {
	byte [] data = new byte[super.aif.AM_SIZE];
	
	data[0] = startConnectivity;
	data[1] = delay;
	data[2] = duration;

	super.aif.sendAM(data,AM_START_EXP_MSG,nodeID);
    }

    /**
     * Starts the experiment on node <nodeID> in <delay> clock ticks and runs 
     * it for <duration> clock ticks
     */
    public void startExperiment(byte delay, byte duration, short nodeID) throws java.io.IOException {
	if(duration == 0) {
	    duration = -1;
	}
	this.startExp((byte) 0,delay,duration,nodeID);
    }

    /**
     * Stops the experiment currently running on node <nodeID> in <delay> 
     * clock ticks
     */
    public void startExperiment(byte delay, short nodeID) throws java.io.IOException {
	this.startExperiment(delay,(byte) -1,nodeID);
    }

    /**
     * Starts TOPOLOGY_QUERY on node <nodeID>
     */
    public void startConnectivity(short nodeID) throws java.io.IOException {
	this.startExp((byte) 1,(byte) 0,(byte) -1,nodeID);
    }

    /**
     * Stops the experiment on node <nodeID> in <delay> clock ticks.  If 
     * <startConnectivity> is 1, it starts topology_query.  If 0, the 
     * loaded experiment is run.
     */
    private void stopExp(byte stopConnectivity, byte delay, short nodeID) throws java.io.IOException {
	byte [] data = new byte[super.aif.AM_SIZE];
	
	data[0] = stopConnectivity;
	data[1] = delay;

	super.aif.sendAM(data,AM_STOP_EXP_MSG,nodeID);
    }

    /**
     * Stops the experiment running on node <nodeID> in <delay> clock ticks
     */
    public void stopExperiment(byte delay, short nodeID) throws java.io.IOException {
	this.stopExp((byte) 0,delay,nodeID);
    }

    /**
     * Stops TOPOLOGY_QUERY on node <nodeID>
     */
    public void stopConnectivity(short nodeID) throws java.io.IOException {
	this.stopExp((byte) 1,(byte) 0,nodeID);
    }

}

File: tinyos-0.4.x/tos/EXPERIMENT.c
/*									tab:4
 * EXPERIMENT.c
 *
 * "Copyright (c) 2000 and The Regents of the University 
 * of California.  All rights reserved.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose, without fee, and without written agreement is
 * hereby granted, provided that the above copyright notice and the following
 * two paragraphs appear in all copies of this software.
 * 
 * IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY FOR
 * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT
 * OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF THE UNIVERSITY OF
 * CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
 * ON AN "AS IS" BASIS, AND THE UNIVERSITY OF CALIFORNIA HAS NO OBLIGATION TO
 * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS."
 *
 * Authors:   Solomon Bien
 * History:   created 8/14/2001
 *
 */

/* This component provides the following capabilities:
   -start an experiment (or topology_query app)
         -specify delay and duration
   -end an experiment (or topology_query app)
         -specify delay
   -handle discover messages--info about a node is sent back in response
   -change the experimental ID (group ID)

   
   All of these things are triggered through AM interfaces.  In order to start
   and stop an experiment, it is necessary to provide mappings for the
   EXPERIMENT_SUB_START_EXPERIMENT() and EXPERIMENT_SUB_STOP_EXPERIMENT()
   functions.  See apps/experiment_test.desc as an example.

   This component uses AM handlers # 89, 90, 91, 92, and 95.
*/

#include "tos.h"
#include "EXPERIMENT.h"

#define MAX_NUM_NEIGHBORS 4  /* must have the same value as the same constant
				in TOPOLOGY_QUERY.c */
#define HIGH_TRANSMIT_POWER 60

typedef struct{
  char exp_id;    // value to which the group ID should be changed
} experiment_msg;

typedef struct{
  short src;          // address of this node
  char pot_setting;   // current POT setting of this node
  //  char naming_scheme;  //include this line if using the NAMING component
  char mote_mode;     // current mode/role of this node
  short neighbors[MAX_NUM_NEIGHBORS];   /* addresses of this node's one-hop 
					   neighbors */
} discover_msg;

typedef struct{
  char start_connectivity;    /* 1 if the start_exp message should start 
				 TOPOLOGY_QUERY, 0 if it should start the 
				 chosen experiment */
  char delay;         // number of clock ticks before experiment should start
  char duration;      // number of clock ticks for which experiment should run
} start_exp_msg;

typedef struct{
  char mode;          // value to which the mode/role should be changed
} mote_mode_msg;

#define MONITOR_MODE     0   // the node is monitoring the experiment
#define EXPERIMENT_MODE  1   // the node is taking part in the experiment

#define TOS_FRAME_TYPE EXPERIMENT_frame
TOS_FRAME_BEGIN(EXPERIMENT_frame) {
  char exp_start_in_x_secs;
  char remaining_running_time;
  char duration_tmp;
  char mode;
  /* uncomment next two lines if signal strength should be increased
     before sending responses to control messages */
  //char pot_stored;
  //char current_pot;
  short respond;
  char pending;
  TOS_Msg data;
}
TOS_FRAME_END(EXPERIMENT_frame);


char TOS_COMMAND(EXPERIMENT_INIT)(){
  TOS_CALL_COMMAND(EXPERIMENT_SUB_INIT)();
  VAR(exp_start_in_x_secs) = -1;
  VAR(remaining_running_time) = -1;
  VAR(mode) = EXPERIMENT_MODE;
  /* uncomment next line if signal strength should be increased
     before sending responses to control messages */
  //VAR(pot_stored) = 0;
  VAR(respond) = 0xffff;
  VAR(pending) = 0;
  return 1;
}

char TOS_COMMAND(EXPERIMENT_START)(){
  TOS_CALL_COMMAND(EXPERIMENT_SUB_START)();
  return 1;
}


char TOS_EVENT(EXPERIMENT_SUB_MSG_SEND_DONE)(TOS_MsgPtr msg) {
  /* uncomment next lines if signal strength should be increased
     before sending responses to control messages */
  //if(VAR(pot_stored) == 1) {
  //  TOS_CALL_COMMAND(EXPERIMENT_POT_SET)(VAR(current_pot));
  //  VAR(pot_stored) = 0;
  //}
  
  if (VAR(pending) && msg == &VAR(data)) {
    VAR(pending) = 0;
    return 1;
  }
  return 0;
}

// handler that changes the group/experimental ID
TOS_MsgPtr TOS_MSG_EVENT(SET_EXP_ID_MSG)(TOS_MsgPtr msg){
  experiment_msg * t;
  
  t = (experiment_msg *) msg->data;
  
  LOCAL_GROUP = t->exp_id;

  return msg;
}

// handler that responds with the current node's info
TOS_MsgPtr TOS_MSG_EVENT(DISCOVER_MSG)(TOS_MsgPtr msg){
  discover_msg * t;
  
  t = (discover_msg *) msg->data;

  TOS_CALL_COMMAND(EXPERIMENT_GREEN_LED_ON)();
  
  if(t->src == 0) {
    VAR(respond) = t->src;
    
    t = (discover_msg *) VAR(data).data;
    
    t->src = TOS_LOCAL_ADDRESS;
    t->pot_setting = TOS_CALL_COMMAND(EXPERIMENT_POT_GET)();
    t->mote_mode = TOS_CALL_COMMAND(EXPERIMENT_GET_MODE)();
    TOS_CALL_COMMAND(EXPERIMENT_YELLOW_LED_ON)();
    // uncomment the next line if using the NAMING component
    //t->naming = TOS_CALL_COMMAND(EXPERIMENT_NAMING_GET_NAMING_SCHEME)();
    /* uncomment next lines if signal strength should be increased
       before sending responses to control messages */
    //VAR(pot_stored) = 1;
    //VAR(current_pot) = TOS_CALL_COMMAND(EXPERIMENT_POT_GET)();
    //TOS_CALL_COMMAND(EXPERIMENT_POT_SET)(HIGH_TRANSMIT_POWER);
    TOS_CALL_COMMAND(EXPERIMENT_GREEN_LED_OFF)();
    TOS_CALL_COMMAND(EXPERIMENT_NEIGHBORS_GET_NEIGHBORS)(t->neighbors,MAX_NUM_NEIGHBORS);
    TOS_CALL_COMMAND(EXPERIMENT_YELLOW_LED_OFF)();
    if(! VAR(pending)) {
      VAR(pending) = 1;
      TOS_CALL_COMMAND(EXPERIMENT_SUB_SEND_MSG)(VAR(respond),AM_MSG(DISCOVER_MSG),&VAR(data));
      TOS_CALL_COMMAND(EXPERIMENT_GREEN_LED_ON)();
    }
  }
  return msg;
}

// handler that starts an experiment
TOS_MsgPtr TOS_MSG_EVENT(START_EXP_MSG)(TOS_MsgPtr msg){
  start_exp_msg * t;

  t = (start_exp_msg *) msg->data;

  if(t->start_connectivity) {
    TOS_CALL_COMMAND(EXPERIMENT_CONNECTIVITY_START)();
  } else {
    if(t->delay == 0) {
      // a mapping for this function must be provided
      TOS_CALL_COMMAND(EXPERIMENT_SUB_START_EXPERIMENT)();
      VAR(remaining_running_time) = t->duration;
    } else {
      VAR(exp_start_in_x_secs) = t->delay;
      VAR(duration_tmp) = t->duration;
    }
  }
  
  return msg;
}


TOS_MsgPtr TOS_MSG_EVENT(STOP_EXP_MSG)(TOS_MsgPtr msg){
  start_exp_msg * t;
  
  t = (start_exp_msg *) msg->data;

  if(t->start_connectivity) {
    TOS_CALL_COMMAND(EXPERIMENT_CONNECTIVITY_STOP)();
  } else {
    if(t->delay == 0) {
      // a mapping for this function must be provided
      TOS_CALL_COMMAND(EXPERIMENT_SUB_STOP_EXPERIMENT)();
    } else {
      VAR(remaining_running_time) = t->delay;
    }
  }
  
  return msg;
}


void TOS_EVENT(EXPERIMENT_CLOCK_EVENT)(){
  if(VAR(remaining_running_time) == 0) {
    TOS_CALL_COMMAND(EXPERIMENT_SUB_STOP_EXPERIMENT)();
  }
  if(VAR(remaining_running_time) >= 0) {
    VAR(remaining_running_time)--;
  }
  
  if(VAR(exp_start_in_x_secs) == 0) {
    TOS_CALL_COMMAND(EXPERIMENT_SUB_START_EXPERIMENT)();
    VAR(remaining_running_time) = VAR(duration_tmp);
  }
  if(VAR(exp_start_in_x_secs) >= 0) {
    VAR(exp_start_in_x_secs)--;
  }
}

char TOS_COMMAND(EXPERIMENT_GET_MODE)() {
  return VAR(mode);
}

// handler that changes the mode/role of the current node
TOS_MsgPtr TOS_MSG_EVENT(MOTE_MODE_MSG)(TOS_MsgPtr msg){
  mote_mode_msg * t;
    
  t = (mote_mode_msg *) msg->data;

  VAR(mode) = t->mode;

  return msg;
}
